pallets/proofs-dealer/src/utils.rs

```diff
@@ -49,22 +88,134 @@ where
+            .ok_or(DispatchError::Arithmetic(ArithmeticError::Overflow))?;
+
+        // Check that the challenges block is lower than the current block.
+        ensure!(
+            challenges_block < frame_system::Pallet::<T>::block_number(),
+            Error::<T>::ChallengesBlockNotReached
+        );
+
+        // Check that the challenges block is greater than current block minus `ChallengeHistoryLength`,
+        // i.e. that the challenges block is within the blocks this pallet keeps track of.
+        expect_or_err!(
+            challenges_block
+                > frame_system::Pallet::<T>::block_number()
+                    .saturating_sub(ChallengeHistoryLengthFor::<T>::get()),
+            "Challenges block is too old, beyond the history this pallet keeps track of. This should not be possible.",
+            Error::<T>::ChallengesBlockTooOld,
+            bool
+        );
+
+        // Get seed for challenges block.
+        let seed = expect_or_err!(
+            BlockToChallengesSeed::<T>::get(challenges_block),
+            "Seed for challenges block not found, when checked it should be within history.",
+            Error::<T>::SeedNotFound
+        );
+
+        // Generate forest challenges from seed.
+        let mut challenges = Self::generate_challenges_from_seed(
+            seed,
+            submitter,
+            RandomChallengesPerBlockFor::<T>::get(),
+        );
+
+        // Check if there's been a Checkpoint Challenge block in between the last block proven and
+        // the current block. If there has been, the Provider should have included proofs for the
+        // challenges in that block.
+        let last_checkpoint_block = LastCheckpointBlock::<T>::get();
+        if last_block_proven < last_checkpoint_block {
+            // Add challenges from the Checkpoint Challenge block.
+            let checkpoint_challenges =
+                expect_or_err!(
+                    BlockToCheckpointChallenges::<T>::get(last_checkpoint_block),
+                    "Checkpoint challenges not found, when dereferencing in last registered checkpoint challenge block.",
+                    Error::<T>::CheckpointChallengesNotFound
+                );
+            challenges.extend(checkpoint_challenges);
+        }
+
+        // Verify forest proof.
+        let forest_keys_proven =
+            ForestVerifierFor::<T>::verify_proof(&root, &challenges, forest_proof)
+                .map_err(|_| Error::<T>::ForestProofVerificationFailed)?;
+
+        // Verify each key proof.
+        for key_proven in forest_keys_proven {
+            // Check that there is a key proof for each key proven.
+            let key_proof = key_proofs
+                .get(&key_proven)
+                .ok_or(Error::<T>::KeyProofNotFound)?;
+
+            // Generate the challenges for the key.
+            let challenges =
+                Self::generate_challenges_from_seed(seed, submitter, key_proof.challenge_count);
```

Comment:
Where does the `challenge_count` come from?

Reply:
Nice catch! My bad. The `challenges_count` should actually be a function of the file size. I'll fix it right away.

Reply:
Actually, I was wrong, this has to be here. In the `challenges_count` field, the Storage Provider specifies how many chunks are proven in this proof. That many chunks should be related to the size of the file, but the Proofs Dealer pallet is abstracted over the `KeyVerifierProofsFor<T>` type, and therefore is not aware of the `size` of files being proven. In fact it is not even aware that what's being proved are files. So this variable is used to generate the challenges, and then it is up to the `KeyVerifier` to check that the `challenges_count` is accurate to the `size` of the file.